File: build_support/model_catalog_validation.rs
use std::collections::{HashMap, HashSet};

#[path = "model_catalog_validation/presentation.rs"]
mod presentation;
#[path = "model_catalog_validation/vision.rs"]
mod vision;

pub(super) fn validate(manifest: &serde_json::Value) {
    assert_eq!(
        manifest
            .get("schema_version")
            .and_then(serde_json::Value::as_u64),
        Some(8),
        "catalog schema_version must be 8"
    );
    assert!(
        manifest.get("model_id_migrations").is_none(),
        "permanent model ID migrations are forbidden"
    );

    let models = array(manifest, "models");
    let profiles = object(manifest, "model_profiles");
    let presentation_variants = object(manifest, "presentation_variants");
    presentation::validate_variants(models, presentation_variants);
    let mut ids = HashSet::new();
    let mut used_profiles = HashSet::new();
    let mut localized_names = HashMap::new();
    for item in models {
        let model = item.as_object().expect("model entries must be objects");
        let id = string(model, "id");
        assert!(ids.insert(id), "duplicate model id {id:?}");
        validate_model_id(id);
        let provider = string(model, "provider");
        assert!(
            id.starts_with(provider_id_prefix(provider)),
            "model ID {id:?} does not match provider {provider:?}"
        );
        let full_name = string(model, "full_name");
        let profile_key = format!("{provider}:{full_name}");
        let profile = profiles
            .get(&profile_key)
            .and_then(serde_json::Value::as_object)
            .unwrap_or_else(|| panic!("missing model profile for {profile_key:?}"));
        used_profiles.insert(profile_key.clone());
        validate_presentation(model, profile);
        validate_reasoning_policy(id, provider, profile);
        for language in ["name_vi", "name_ko", "name_en"] {
            let name =
                presentation::localized_name(model, profile, presentation_variants, language);
            let prefix = provider_prefix(string(model, "provider"));
            assert!(
                name.starts_with(&format!("{prefix} ")),
                "{language} for {id:?} must start with {prefix:?}"
            );
            validate_localized_name(language, id, &name);
            if let Some(existing_full_name) =
                localized_names.insert((language, prefix, name.clone()), profile_key.clone())
            {
                assert_eq!(
                    existing_full_name, profile_key,
                    "distinct API models share {language} name within provider group: {name:?}"
                );
            }
        }
    }
    assert_eq!(
        used_profiles.len(),
        profiles.len(),
        "model_profiles contains an unreferenced API model"
    );
    vision::validate_request_profiles(manifest, models, profiles);

    let enabled_ids: HashSet<&str> = models
        .iter()
        .filter_map(serde_json::Value::as_object)
        .filter(|model| model.get("enabled").and_then(serde_json::Value::as_bool) == Some(true))
        .map(|model| string(model, "id"))
        .collect();
    validate_chains(manifest, &enabled_ids);
    validate_endpoints(manifest, models);
}

fn validate_model_id(id: &str) {
    assert!(
        id.bytes()
            .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'-')
            && !id.starts_with('-')
            && !id.ends_with('-')
            && !id.contains("--"),
        "model ID must be lowercase ASCII kebab-case: {id:?}"
    );
    let segments: Vec<&str> = id.split('-').collect();
    assert!(
        [
            "google",
            "groq",
            "openrouter",
            "taalas",
            "qrserver",
            "local",
        ]
        .contains(&segments[0]),
        "model ID must start with a catalog provider: {id:?}"
    );
    assert!(
        ["text", "vision", "audio", "search"].contains(segments.last().unwrap()),
        "model ID must end with a capability: {id:?}"
    );
    for lifecycle in [
        "preview",
        "latest",
        "experimental",
        "stable",
        "deprecated",
        "retired",
    ] {
        assert!(
            !segments.contains(&lifecycle),
            "model ID contains mutable lifecycle word {lifecycle:?}: {id:?}"
        );
    }
}

fn validate_presentation(
    model: &serde_json::Map<String, serde_json::Value>,
    profile: &serde_json::Map<String, serde_json::Value>,
) {
    let id = string(model, "id");
    for field in [
        "name_vi",
        "name_ko",
        "name_en",
        "quota_vi",
        "quota_ko",
        "quota_en",
        "supports_search",
        "search_tool_enabled_by_default",
        "intelligence_tier",
        "reasoning_policy",
    ] {
        assert!(
            model.get(field).is_none(),
            "model row {id:?} duplicates profile-owned field {field:?}"
        );
    }
    assert!(
        profile
            .get("intelligence_tier")
            .and_then(serde_json::Value::as_u64)
            .is_some_and(|value| (1..=6).contains(&value)),
        "intelligence_tier for {id:?} must be 1..=6"
    );
    assert!(
        model
            .get("typical_latency_ms")
            .and_then(serde_json::Value::as_u64)
            .is_some_and(|value| (1..=i32::MAX as u64).contains(&value)),
        "typical_latency_ms for {id:?} must be a positive cross-platform i32"
    );
    assert!(
        !string(model, "performance_source").trim().is_empty(),
        "performance_source for {id:?} must not be empty"
    );
    assert!(
        profile
            .get("supports_search")
            .and_then(serde_json::Value::as_bool)
            .is_some(),
        "supports_search for {id:?} must be boolean"
    );
    let search_tool_enabled_by_default = profile
        .get("search_tool_enabled_by_default")
        .and_then(serde_json::Value::as_bool)
        .unwrap_or_else(|| panic!("search_tool_enabled_by_default for {id:?} must be boolean"));
    assert!(
        !search_tool_enabled_by_default
            || profile
                .get("supports_search")
                .and_then(serde_json::Value::as_bool)
                == Some(true),
        "search_tool_enabled_by_default for {id:?} requires supports_search"
    );
    validate_quota_triplet(id, profile);
}

fn validate_reasoning_policy(
    id: &str,
    provider: &str,
    profile: &serde_json::Map<String, serde_json::Value>,
) {
    let policy = string(profile, "reasoning_policy");
    assert!(
        [
            "not-applicable",
            "gemini-disabled",
            "gemini-minimal",
            "openai-none",
            "openai-low",
            "provider-managed",
            "live-profile",
        ]
        .contains(&policy),
        "unsupported reasoning_policy for {id:?}: {policy:?}"
    );
    let compatible = match provider {
        "google" => matches!(policy, "gemini-disabled" | "gemini-minimal"),
        "gemini-live" => policy == "live-profile",
        "groq" | "openrouter" => matches!(
            policy,
            "not-applicable" | "openai-none" | "openai-low" | "provider-managed"
        ),
        _ => policy == "not-applicable",
    };
    assert!(
        compatible,
        "reasoning_policy {policy:?} is incompatible with provider {provider:?} for {id:?}"
    );
}

fn validate_localized_name(language: &str, id: &str, name: &str) {
    let lower = name.to_lowercase();
    let forbidden: &[&str] = match language {
        "name_vi" => &[
            " ảnh",
            " chữ",
            "ocr",
            "định vị",
            "giới hạn",
            "sắp dừng",
            "dài dòng",
            "thử nghiệm",
            "suy luận",
            "dịch máy",
            "live lỗi",
        ],
        "name_ko" => &[
            "비전",
            "텍스트",
            "ocr",
            "위치 찾기",
            "제한",
            "곧 종료",
            "장황",
            "실험",
            "추론",
            "기계 번역",
            "오류",
        ],
        "name_en" => &[
            " vision",
            " text",
            "ocr",
            "grounding",
            "limited",
            "retiring",
            "verbose",
            "experimental",
            "reasoning",
            "machine translation",
            "live errors",
        ],
        _ => unreachable!(),
    };
    assert!(
        !forbidden.iter().any(|term| lower.contains(term)),
        "{language} for {id:?} contains forbidden category/lifecycle wording: {name:?}"
    );
}

fn validate_quota_triplet(id: &str, profile: &serde_json::Map<String, serde_json::Value>) {
    let vi = string(profile, "quota_vi");
    let ko = string(profile, "quota_ko");
    let en = string(profile, "quota_en");
    if (vi, ko, en) == ("Không giới hạn", "무제한", "Unlimited") {
        return;
    }
    let vi_count = quota_count(vi, " lượt/ngày");
    let ko_count = quota_count(ko, "회/일");
    let en_count = quota_count(en, " requests/day");
    assert!(
        vi_count.is_some() && vi_count == ko_count && ko_count == en_count,
        "quota labels for {id:?} must be the same positive daily request count or Unlimited"
    );
}

fn quota_count(label: &str, suffix: &str) -> Option<u64> {
    label
        .strip_suffix(suffix)?
        .parse::<u64>()
        .ok()
        .filter(|count| *count > 0)
}

fn validate_chains(manifest: &serde_json::Value, enabled_ids: &HashSet<&str>) {
    let priority = object(manifest, "priority_chains");
    for key in ["image_to_text", "text_to_text"] {
        validate_chain(array_from(priority, key), key, enabled_ids, None);
    }

    let features = object(manifest, "feature_model_chains");
    for key in ["help_assistant", "computer_control_grounding"] {
        validate_chain(array_from(features, key), key, enabled_ids, Some(2));
    }

    let constants = object(manifest, "constants");
    assert_eq!(
        string(constants, "default_image_model_id"),
        array_from(priority, "image_to_text")[0].as_str().unwrap(),
        "default image model must lead image_to_text"
    );
    assert_eq!(
        string(constants, "default_text_model_id"),
        array_from(priority, "text_to_text")[0].as_str().unwrap(),
        "default text model must lead text_to_text"
    );
}

fn validate_chain(
    chain: &[serde_json::Value],
    key: &str,
    enabled_ids: &HashSet<&str>,
    required_len: Option<usize>,
) {
    if let Some(required_len) = required_len {
        assert_eq!(
            chain.len(),
            required_len,
            "{key} must define primary and fallback models"
        );
    }
    assert!(!chain.is_empty(), "{key} must not be empty");
    let mut seen = HashSet::new();
    for value in chain {
        let id = value
            .as_str()
            .unwrap_or_else(|| panic!("{key} model IDs must be strings"));
        assert!(
            enabled_ids.contains(id),
            "{key} references disabled or unknown model {id:?}"
        );
        assert!(seen.insert(id), "{key} contains duplicate model {id:?}");
    }
}

fn validate_endpoints(manifest: &serde_json::Value, models: &[serde_json::Value]) {
    let endpoints = object(manifest, "endpoints");
    for (endpoint, value) in endpoints {
        let metadata = value
            .as_object()
            .expect("endpoint lifecycle must be an object");
        let lifecycle = string(metadata, "lifecycle");
        assert!(
            ["stable", "preview", "experimental", "deprecated", "retired"].contains(&lifecycle),
            "invalid lifecycle for {endpoint}"
        );
        assert!(
            !string(metadata, "verified_at").is_empty(),
            "missing verified_at for {endpoint}"
        );
        if let Some(replacement) = metadata
            .get("replacement")
            .and_then(serde_json::Value::as_str)
        {
            assert!(
                endpoints.contains_key(replacement),
                "unknown replacement {replacement:?}"
            );
        }
        validate_live_profile(endpoint, metadata);
    }

    let forbidden = |endpoint: &str| {
        endpoints
            .get(endpoint)
            .and_then(|value| value.get("lifecycle"))
            .and_then(serde_json::Value::as_str)
            .is_some_and(|stage| matches!(stage, "deprecated" | "retired"))
    };
    let constants = object(manifest, "constants");
    for key in ["gemini_live_api_model_2_5", "gemini_live_api_model_3_1"] {
        let endpoint = string(constants, key);
        let profile = endpoints
            .get(endpoint)
            .and_then(serde_json::Value::as_object)
            .unwrap_or_else(|| panic!("{key} must reference a catalog endpoint"));
        assert_eq!(
            profile
                .get("live_protocol")
                .and_then(serde_json::Value::as_str),
            Some("native-audio"),
            "{key} endpoint must use the native-audio protocol"
        );
    }
    for model in models.iter().filter_map(serde_json::Value::as_object) {
        if model.get("enabled").and_then(serde_json::Value::as_bool) == Some(true) {
            let endpoint = string(model, "full_name");
            assert!(
                !forbidden(endpoint),
                "enabled model uses deprecated/retired endpoint {endpoint:?}"
            );
        }
    }
    let defaults = object(manifest, "defaults");
    assert!(
        !forbidden(string(defaults, "tts_gemini_live_model")),
        "default TTS endpoint is deprecated/retired"
    );
    for item in array(manifest, "tts_gemini_models") {
        let endpoint = string(item.as_object().unwrap(), "api_model");
        assert!(
            !forbidden(endpoint),
            "TTS option is deprecated/retired: {endpoint}"
        );
    }
}

fn validate_live_profile(endpoint: &str, metadata: &serde_json::Map<String, serde_json::Value>) {
    if let Some(thinking) = metadata.get("live_thinking") {
        let thinking = thinking
            .as_object()
            .unwrap_or_else(|| panic!("live_thinking for {endpoint} must be an object"));
        let kind = string(thinking, "kind");
        let value = thinking
            .get("value")
            .unwrap_or_else(|| panic!("live_thinking value is required for {endpoint}"));
        match kind {
            "budget" => assert!(
                value
                    .as_u64()
                    .is_some_and(|budget| budget <= i32::MAX as u64),
                "live_thinking budget for {endpoint} must be a non-negative 32-bit integer"
            ),
            "level" => assert!(
                value.as_str().is_some_and(|level| !level.trim().is_empty()),
                "live_thinking level for {endpoint} must be a non-empty string"
            ),
            _ => panic!("unsupported live_thinking kind {kind:?} for {endpoint}"),
        }
    }
    if let Some(limit) = metadata.get("live_max_output_tokens") {
        assert!(
            limit
                .as_u64()
                .is_some_and(|limit| (1..=u32::MAX as u64).contains(&limit)),
            "live_max_output_tokens for {endpoint} must be a positive u32"
        );
    }
    if let Some(value) = metadata.get("live_automatic_activity_detection_default") {
        assert!(
            value.is_boolean(),
            "live_automatic_activity_detection_default for {endpoint} must be boolean"
        );
    }
    if let Some(value) = metadata.get("live_protocol") {
        assert!(
            value.as_str().is_some_and(|value| !value.trim().is_empty()),
            "live_protocol for {endpoint} must be a non-empty string"
        );
    }
    if metadata
        .get("live_protocol")
        .and_then(serde_json::Value::as_str)
        == Some("native-audio")
    {
        assert!(
            metadata.get("live_thinking").is_some(),
            "native-audio endpoint {endpoint} must define live_thinking"
        );
        assert!(
            metadata.get("live_max_output_tokens").is_some(),
            "native-audio endpoint {endpoint} must define live_max_output_tokens"
        );
    }
}

fn provider_prefix(provider: &str) -> &str {
    match provider {
        "google" | "google-gtx" | "gemini-live" => "GG",
        "groq" => "G",
        "openrouter" => "O",
        "taalas" => "T",
        "parakeet" | "qwen3" => "L",
        "qrserver" => "QR",
        _ => panic!("provider {provider:?} has no localized-name prefix"),
    }
}

fn provider_id_prefix(provider: &str) -> &str {
    match provider {
        "google" | "google-gtx" | "gemini-live" => "google-",
        "groq" => "groq-",
        "openrouter" => "openrouter-",
        "taalas" => "taalas-",
        "parakeet" | "qwen3" => "local-",
        "qrserver" => "qrserver-",
        _ => panic!("provider {provider:?} has no internal-ID prefix"),
    }
}

fn object<'a>(
    value: &'a serde_json::Value,
    key: &str,
) -> &'a serde_json::Map<String, serde_json::Value> {
    value
        .get(key)
        .and_then(serde_json::Value::as_object)
        .unwrap_or_else(|| panic!("manifest key {key:?} must be an object"))
}

fn array<'a>(value: &'a serde_json::Value, key: &str) -> &'a Vec<serde_json::Value> {
    value
        .get(key)
        .and_then(serde_json::Value::as_array)
        .unwrap_or_else(|| panic!("manifest key {key:?} must be an array"))
}

fn array_from<'a>(
    value: &'a serde_json::Map<String, serde_json::Value>,
    key: &str,
) -> &'a Vec<serde_json::Value> {
    value
        .get(key)
        .and_then(serde_json::Value::as_array)
        .unwrap_or_else(|| panic!("manifest object key {key:?} must be an array"))
}

fn string<'a>(value: &'a serde_json::Map<String, serde_json::Value>, key: &str) -> &'a str {
    value
        .get(key)
        .and_then(serde_json::Value::as_str)
        .unwrap_or_else(|| panic!("manifest object key {key:?} must be a string"))
}

File: scripts/model_catalog_validation.py
from __future__ import annotations

import re


PROVIDER_NAME_PREFIXES = {
    "google": "GG",
    "google-gtx": "GG",
    "gemini-live": "GG",
    "groq": "G",
    "openrouter": "O",
    "nvidia": "N",
    "taalas": "T",
    "parakeet": "L",
    "qwen3": "L",
    "qrserver": "QR",
}

PROVIDER_ID_PREFIXES = {
    "google": "google-",
    "google-gtx": "google-",
    "gemini-live": "google-",
    "groq": "groq-",
    "openrouter": "openrouter-",
    "nvidia": "nvidia-",
    "taalas": "taalas-",
    "parakeet": "local-",
    "qwen3": "local-",
    "qrserver": "qrserver-",
}

FORBIDDEN_NAME_TERMS = {
    "vi": (
        " ảnh", " chữ", "ocr", "định vị", "giới hạn", "sắp dừng",
        "dài dòng", "thử nghiệm", "suy luận", "dịch máy", "live lỗi",
    ),
    "ko": (
        "비전", "텍스트", "ocr", "위치 찾기", "제한", "곧 종료",
        "장황", "실험", "추론", "기계 번역", "오류",
    ),
    "en": (
        " vision", " text", "ocr", "grounding", "limited", "retiring",
        "verbose", "experimental", "reasoning", "machine translation",
        "live errors",
    ),
}

PROFILE_FIELDS = (
    "name_vi", "name_ko", "name_en", "quota_vi", "quota_ko", "quota_en",
    "supports_search", "search_tool_enabled_by_default", "intelligence_tier",
    "reasoning_policy",
)


def validate_manifest(manifest: dict) -> None:
    if manifest.get("schema_version") != 8:
        raise ValueError("catalog schema_version must be 8")
    if "model_id_migrations" in manifest:
        raise ValueError("permanent model ID migrations are forbidden")
    models = manifest["models"]
    presentation_variants = manifest["presentation_variants"]
    _validate_presentation_variants(models, presentation_variants)
    _validate_models(models, manifest["model_profiles"], presentation_variants)
    _validate_vision_request_profiles(manifest)
    enabled_ids = {model["id"] for model in models if model["enabled"]}
    _validate_chains(manifest, enabled_ids)
    _validate_endpoints(manifest, models)


def _validate_vision_request_profiles(manifest: dict) -> None:
    expected = {
        f"{model['provider']}:{model['full_name']}"
        for model in manifest["models"]
        if (
            model["enabled"]
            and model["model_type"] == "Vision"
            and model["provider"] in {"google", "groq", "openrouter", "nvidia"}
        )
    }
    request_profiles = manifest.get("vision_request_profiles")
    if not isinstance(request_profiles, dict) or set(request_profiles) != expected:
        raise ValueError(
            "vision_request_profiles must cover every enabled ordinary LLM "
            "vision endpoint exactly"
        )

    for profile_key, request_profile in request_profiles.items():
        if profile_key not in manifest["model_profiles"]:
            raise ValueError(
                f"vision request profile has no model profile: {profile_key}"
            )
        if set(request_profile) != {
            "input_order",
            "media_resolution",
            "sampling",
            "max_output_tokens",
            "structured_output",
        }:
            raise ValueError(
                f"vision request profile fields drifted for {profile_key}"
            )
        if request_profile.get("input_order") not in {"text-first", "image-first"}:
            raise ValueError(f"unsupported input_order for {profile_key}")
        if request_profile.get("media_resolution") != "provider-default":
            raise ValueError(f"unsupported media_resolution for {profile_key}")
        sampling = request_profile.get("sampling")
        if sampling not in {"provider-default", "qwen3-groq-non-thinking"}:
            raise ValueError(f"unsupported sampling policy for {profile_key}")
        max_output_tokens = request_profile.get("max_output_tokens")
        if max_output_tokens is not None and (
            isinstance(max_output_tokens, bool)
            or not isinstance(max_output_tokens, int)
            or not 1 <= max_output_tokens <= 0xFFFF_FFFF
        ):
            raise ValueError(f"invalid max_output_tokens for {profile_key}")
        if request_profile.get("structured_output") not in {
            "unsupported", "prompt-only", "json-object", "strict-json-schema",
        }:
            raise ValueError(f"unsupported structured_output for {profile_key}")
        if sampling == "qwen3-groq-non-thinking" and not (
            profile_key.startswith("groq:")
            and manifest["model_profiles"][profile_key]["reasoning_policy"]
            == "openai-none"
            and max_output_tokens is not None
        ):
            raise ValueError(
                "qwen3-groq-non-thinking requires Groq reasoning policy "
                "openai-none and an output limit"
            )


def _validate_presentation_variants(
    models: list[dict],
    variants: dict[str, dict],
) -> None:
    if not isinstance(variants, dict) or not variants:
        raise ValueError("presentation_variants must be a non-empty object")
    required_fields = {"suffix_vi", "suffix_ko", "suffix_en"}
    for key, variant in variants.items():
        if not isinstance(variant, dict) or set(variant) != required_fields:
            raise ValueError(
                f"presentation variant {key!r} must define only localized suffixes"
            )
        for field in required_fields:
            suffix = variant[field]
            if (
                not isinstance(suffix, str)
                or not suffix.startswith(" (")
                or not suffix.endswith(")")
                or not suffix[2:-1]
                or suffix[2:-1].strip() != suffix[2:-1]
            ):
                raise ValueError(
                    f"{field} for presentation variant {key!r} must be "
                    "a parenthesized suffix"
                )

    used: set[str] = set()
    for model in models:
        key = model.get("presentation_variant")
        if key is None:
            continue
        if not isinstance(key, str) or key not in variants:
            raise ValueError(
                f"unknown presentation variant {key!r} for {model['id']}"
            )
        used.add(key)
        sibling_fields = ("provider", "full_name", "model_type")
        siblings = sum(
            all(candidate.get(field) == model.get(field) for field in sibling_fields)
            for candidate in models
        )
        if siblings <= 1:
            raise ValueError(
                f"presentation variant on {model['id']} requires a behavioral "
                "sibling using the same endpoint"
            )
    if used != set(variants):
        raise ValueError("presentation_variants contains an unreferenced variant")


def _localized_model_name(
    model: dict,
    profile: dict,
    variants: dict[str, dict],
    language: str,
) -> str:
    name = profile[f"name_{language}"]
    key = model.get("presentation_variant")
    if key is not None:
        name += variants[key][f"suffix_{language}"]
    return name


def _validate_models(
    models: list[dict],
    profiles: dict[str, dict],
    presentation_variants: dict[str, dict],
) -> None:
    ids: set[str] = set()
    localized_names: dict[tuple[str, str, str], str] = {}
    used_profiles: set[str] = set()
    for model in models:
        model_id = model["id"]
        if model_id in ids:
            raise ValueError(f"duplicate model id: {model_id}")
        ids.add(model_id)
        _validate_model_id(model)
        full_name = model["full_name"]
        profile_key = f"{model['provider']}:{full_name}"
        profile = profiles.get(profile_key)
        if not isinstance(profile, dict):
            raise ValueError(f"missing model profile for {profile_key}")
        used_profiles.add(profile_key)
        _validate_presentation(model, profile)
        _validate_reasoning_policy(model_id, model["provider"], profile)
        prefix = PROVIDER_NAME_PREFIXES.get(model["provider"])
        if prefix is None:
            raise ValueError(f"missing localized-name prefix for {model['provider']}")
        for language in ("vi", "ko", "en"):
            name = _localized_model_name(
                model,
                profile,
                presentation_variants,
                language,
            )
            if not name.startswith(f"{prefix} "):
                raise ValueError(
                    f"{language} name for {model_id} must start with {prefix}"
                )
            _validate_localized_name(language, model_id, name)
            key = (language, prefix, name)
            existing = localized_names.setdefault(key, profile_key)
            if existing != profile_key:
                raise ValueError(
                    f"distinct API models share {language} name in {prefix}: {name}"
                )
    if used_profiles != set(profiles):
        raise ValueError("model_profiles contains an unreferenced API model")


def _validate_model_id(model: dict) -> None:
    model_id = model["id"]
    segments = model_id.split("-")
    lifecycle_words = {
        "preview", "latest", "experimental", "stable", "deprecated", "retired",
    }
    if (
        re.fullmatch(r"[a-z0-9]+(?:-[a-z0-9]+)*", model_id) is None
        or segments[0] not in {
            "google", "groq", "openrouter", "nvidia", "taalas", "qrserver", "local",
        }
        or segments[-1] not in {"text", "vision", "audio", "search"}
        or lifecycle_words.intersection(segments)
    ):
        raise ValueError(f"invalid durable model id: {model_id}")
    prefix = PROVIDER_ID_PREFIXES.get(model["provider"])
    if prefix is None or not model_id.startswith(prefix):
        raise ValueError(
            f"model id {model_id} does not match provider {model['provider']}"
        )


def _validate_presentation(model: dict, profile: dict) -> None:
    model_id = model["id"]
    duplicated = set(PROFILE_FIELDS).intersection(model)
    if duplicated:
        raise ValueError(
            f"model row {model_id} duplicates profile fields: {sorted(duplicated)}"
        )
    intelligence = profile.get("intelligence_tier")
    latency = model.get("typical_latency_ms")
    if (
        isinstance(intelligence, bool)
        or not isinstance(intelligence, int)
        or not 1 <= intelligence <= 6
    ):
        raise ValueError(f"intelligence_tier for {model_id} must be 1..6")
    if (
        isinstance(latency, bool)
        or not isinstance(latency, int)
        or not 1 <= latency <= 2_147_483_647
    ):
        raise ValueError(
            f"typical_latency_ms for {model_id} must be a positive cross-platform i32"
        )
    if not isinstance(model.get("performance_source"), str) or not model["performance_source"].strip():
        raise ValueError(f"performance_source for {model_id} must not be empty")
    if not isinstance(profile.get("supports_search"), bool):
        raise ValueError(f"supports_search for {model_id} must be boolean")
    search_enabled = profile.get("search_tool_enabled_by_default")
    if not isinstance(search_enabled, bool):
        raise ValueError(
            f"search_tool_enabled_by_default for {model_id} must be boolean"
        )
    if search_enabled and not profile["supports_search"]:
        raise ValueError(
            f"search_tool_enabled_by_default for {model_id} requires supports_search"
        )
    _validate_quota(model_id, profile)


def _validate_reasoning_policy(model_id: str, provider: str, profile: dict) -> None:
    policy = profile.get("reasoning_policy")
    allowed = {
        "not-applicable",
        "gemini-disabled",
        "gemini-minimal",
        "gemini-low",
        "openai-none",
        "openai-low",
        "provider-managed",
        "live-profile",
    }
    if policy not in allowed:
        raise ValueError(
            f"unsupported reasoning_policy for {model_id!r}: {policy!r}"
        )
    if provider == "google":
        compatible = policy in {"gemini-disabled", "gemini-minimal", "gemini-low"}
    elif provider == "gemini-live":
        compatible = policy == "live-profile"
    elif provider in {"groq", "openrouter", "nvidia"}:
        compatible = policy in {
            "not-applicable",
            "openai-none",
            "openai-low",
            "provider-managed",
        }
    else:
        compatible = policy == "not-applicable"
    if not compatible:
        raise ValueError(
            f"reasoning_policy {policy!r} is incompatible with provider "
            f"{provider!r} for {model_id!r}"
        )


def _validate_localized_name(language: str, model_id: str, name: str) -> None:
    lower = name.lower()
    if any(term in lower for term in FORBIDDEN_NAME_TERMS[language]):
        raise ValueError(
            f"{language} name for {model_id} has forbidden category/lifecycle wording: {name}"
        )


def _validate_quota(model_id: str, profile: dict) -> None:
    labels = (profile["quota_vi"], profile["quota_ko"], profile["quota_en"])
    if labels == ("Không giới hạn", "무제한", "Unlimited"):
        return
    suffixes = (" lượt/ngày", "회/일", " requests/day")
    counts = []
    for label, suffix in zip(labels, suffixes, strict=True):
        if not label.endswith(suffix) or not label.removesuffix(suffix).isdigit():
            raise ValueError(f"invalid daily quota wording for {model_id}: {label}")
        counts.append(int(label.removesuffix(suffix)))
    if len(set(counts)) != 1 or counts[0] <= 0:
        raise ValueError(f"quota counts disagree for {model_id}")


def _validate_chains(manifest: dict, enabled_ids: set[str]) -> None:
    priority_chains = manifest["priority_chains"]
    for key in ("image_to_text", "text_to_text"):
        _validate_chain(priority_chains.get(key), key, enabled_ids)
    constants = manifest["constants"]
    if constants["default_image_model_id"] != priority_chains["image_to_text"][0]:
        raise ValueError("default image model must lead image_to_text")
    if constants["default_text_model_id"] != priority_chains["text_to_text"][0]:
        raise ValueError("default text model must lead text_to_text")
    for key in ("help_assistant", "computer_control_grounding"):
        chain = manifest["feature_model_chains"].get(key)
        _validate_chain(chain, key, enabled_ids)
        if len(chain) != 2:
            raise ValueError(f"{key} must define primary and fallback models")


def _validate_chain(chain: list[str], key: str, enabled_ids: set[str]) -> None:
    if not isinstance(chain, list) or not chain or len(chain) != len(set(chain)):
        raise ValueError(f"{key} must be a non-empty unique model chain")
    unknown = [model_id for model_id in chain if model_id not in enabled_ids]
    if unknown:
        raise ValueError(f"{key} references disabled or unknown models: {unknown}")


def _validate_endpoints(manifest: dict, models: list[dict]) -> None:
    endpoints = manifest["endpoints"]
    allowed = {"stable", "preview", "experimental", "deprecated", "retired"}
    for endpoint, metadata in endpoints.items():
        if metadata.get("lifecycle") not in allowed or not metadata.get("verified_at"):
            raise ValueError(f"invalid lifecycle metadata for {endpoint}")
        replacement = metadata.get("replacement")
        if replacement is not None and replacement not in endpoints:
            raise ValueError(f"unknown endpoint replacement: {replacement}")
        _validate_live_profile(endpoint, metadata)
    forbidden = {
        endpoint for endpoint, metadata in endpoints.items()
        if metadata["lifecycle"] in {"deprecated", "retired"}
    }
    for key in ("gemini_live_api_model_2_5", "gemini_live_api_model_3_1"):
        endpoint = manifest["constants"][key]
        profile = endpoints.get(endpoint)
        if profile is None or profile.get("live_protocol") != "native-audio":
            raise ValueError(f"{key} must reference a native-audio endpoint")
    if any(model["enabled"] and model["full_name"] in forbidden for model in models):
        raise ValueError("enabled model uses a retired endpoint")
    runtime = [manifest["defaults"]["tts_gemini_live_model"]]
    runtime += [item["api_model"] for item in manifest["tts_gemini_models"]]
    if forbidden.intersection(runtime):
        raise ValueError("deprecated/retired endpoint cannot be a runtime default")


def _validate_live_profile(endpoint: str, metadata: dict) -> None:
    thinking = metadata.get("live_thinking")
    if thinking is not None:
        if not isinstance(thinking, dict):
            raise ValueError(f"live_thinking for {endpoint} must be an object")
        kind, value = thinking.get("kind"), thinking.get("value")
        valid = (
            kind == "budget"
            and isinstance(value, int)
            and not isinstance(value, bool)
            and 0 <= value <= 2_147_483_647
        ) or (kind == "level" and isinstance(value, str) and bool(value.strip()))
        if not valid:
            raise ValueError(f"invalid live_thinking for {endpoint}")
    limit = metadata.get("live_max_output_tokens")
    if limit is not None and (
        isinstance(limit, bool) or not isinstance(limit, int) or not 1 <= limit <= 0xFFFF_FFFF
    ):
        raise ValueError(f"invalid live_max_output_tokens for {endpoint}")
    automatic = metadata.get("live_automatic_activity_detection_default")
    if automatic is not None and not isinstance(automatic, bool):
        raise ValueError(f"invalid activity detection default for {endpoint}")
    protocol = metadata.get("live_protocol")
    if protocol is not None and (not isinstance(protocol, str) or not protocol.strip()):
        raise ValueError(f"invalid live_protocol for {endpoint}")
    if protocol == "native-audio" and (thinking is None or limit is None):
        raise ValueError(
            f"native-audio endpoint {endpoint} must define thinking and output policy"
        )
